### PR TITLE
Non-record: CoDA-GQA Differential Attention — First Differential Attention Submission (val_bpb=1.1580)

### PR DESCRIPTION
**Pull request by anthony-maio**
## Non-Record Submission: CoDA-GQA Differential Attention

First application of **differential attention** to Parameter Golf. CoDA-GQA (Maio, 2026) sharpens attention by subtracting a gated inhibitory noise stream, where the noise query is produced via learnable orthogonal rotation — eliminating the need for a second W_q matrix.

### Controlled Ablation (8×H100 SXM, 600s, seed=1337)

| | Baseline (CoDA OFF) | CoDA ON | Delta |
|---|---|---|---|
| step_avg | 104.7ms | 136.3ms | +30.1% |
| Steps | 5,732 | 4,404 | -23.2% |
| **Sliding window bpb** | **1.1459** | **1.1580** | **+0.0121** |

### Finding

CoDA trains stably from scratch without model collapse (lambda gates activate smoothly from sigmoid(-6) ≈ 0.0025). However, the 30% per-step overhead from dual SDPA outweighs the noise-cancellation benefit within a 600-second budget, resulting in 0.012 bpb worse than baseline.

The architecture likely needs the unlimited compute track (4+ hours) where the differential mechanism has time to fully activate — consistent with the original CoDA-GQA-L paper's 2-phase training protocol (2000 steps differential + 600 steps memory).

### How CoDA Works
```
q_noise = PairwiseRotate(q, theta)          # orthogonal rotation (~0 params)
out_sig = Attn(q, k, v)                    # signal (same KV)
out_noise = Attn(q_noise, k, v)            # noise (same KV)
lambda = sigmoid(Linear(x; bias=-6))       # input-dependent gate
out = RMSNorm(out_sig - lambda * out_noise)
```
Parameter overhead: ~48K params (0.2% of 27M model). No second W_q matrix needed.

### Credits
- CoDA-GQA-L: Maio, 2026 (Zenodo DOI: 10.5281/zenodo.18804610)
- Differential attention concept: Ye et al., 2024 (arXiv:2410.05258)
- Base model: PR #414 by @signalrush

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: CoDA-GQA Differential Attention — First Differential Attention Submission (val_bpb=1.1580)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

--- ## Analysis ### ILLEGAL n-gram family bug (target XOR'd into context hash key) NOT present. The n-gram cache (`NgramBackoffCache`, lines 944–988) has two hash functions: - `_hash_ctx` (lines 957–961): hashes only the context tokens, no target involved. - `_hash_full` (lines 962–963): XORs `ctx_hash ^ (target * PRIME)` to build a (context, target) pair hash. This is the correct, standard count-min-sketch bigram/n-gram design: `ctx_tables` counts context occurrences and `full_tables` counts (context, target) co-occurrences. The target appears in `_hash_full` legitimately to distinguish different successor tokens for the same context. The ILLEGAL pattern would be if the target leaked into the context-only hash (`_hash_ctx`) — it does not. ### ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first) NOT present. There are zero gradient updates on `val_tokens` at any point in the script. All evaluation functions (`eval_val`, `eval_val_sliding`, `eval_val_ngram_backoff`) run under `torch.inference_mode()` (lines 903, 1016). The training loop only trains on `train_loader` data. ### LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard) NOT applicable. There is no TTT in this submission. No test-time adaptation, no val-token fine-tuning. ### HOLD scored-region SLOT NOT applicable. The n-gram eval (`eval_val_ngram_backoff`) uses a strict score-first update discipline: all tokens in each sliding window are scored (lines 1043–1064), and `cache.update(...)` is called only after the inner scoring loop exits (lines 1065–1068), using `scored_up_to` to prevent any token from appearing in the cache before it is scored. The cache is seeded exclusively from prior windows. ### Architecture / novel technique This PR introduces CoDA-GQA Differential Attention (lines 486–575, 701–774):...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
